### PR TITLE
[Flax] Add Vae for Stable Diffusion

### PR DESCRIPTION
**Pull request by patrickvonplaten**

### CODE REVIEW

File: src/diffusers/modeling_flax_utils.py [hunk 1]
```diff
@@ -294,6 +294,7 @@ def from_pretrained(
             local_files_only=local_files_only,
             use_auth_token=use_auth_token,
             revision=revision,
+            subfolder=subfolder,
```
Comment (patrickvonplaten) [inline src/diffusers/modeling_flax_utils.py:297]: Let's allow downloading from subfolders in Flax as well

**Comment by patrickvonplaten** [comment]
Will add tests and docstring later:

To verify that it works:

```python
#!/usr/bin/env python3
from diffusers import FlaxAutoencoderKL, AutoencoderKL
import torch
import numpy as np
import jax.numpy as jnp

model_flax, params = FlaxAutoencoderKL.from_pretrained("fusing/sd-v1-4-flax", subfolder="vae", use_auth_token=True)
model_pt = AutoencoderKL.from_pretrained("CompVis/stable-diffusion-v1-4", subfolder="vae", use_auth_token=True)

sample_shape = (1, model_flax.config.in_channels, model_flax.config.sample_size, model_flax.config.sample_size)
sample = jnp.ones(sample_shape, dtype=jnp.float32)

sample_out_flax = model_flax.apply({"params": params}, sample)

sample_pt = torch.from_numpy(np.asarray(sample))

with torch.no_grad():
    sample_out_pt = model_pt(sample_pt)

max_diff = np.max(np.abs(np.asarray(sample_out_flax.sample) - sample_out_pt.sample.numpy()))
print("Max diff", max_diff)
```

File: src/diffusers/models/vae_flax.py [hunk 2]
```diff
@@ -0,0 +1,610 @@
...
+
+        sample = self.conv_norm_out(sample)
+        sample = nn.swish(sample)
+        sample = self.conv_out(sample)
+
+        return sample
+
+
+class DiagonalGaussianDistribution(object):
+    # TODO: should we pass dtype?
```
Comment (pcuenca) [inline src/diffusers/models/vae_flax.py:474]: Maybe remove this comment (I think it was originally mine :) )

File: src/diffusers/models/vae_flax.py [hunk 3]
```diff
@@ -0,0 +1,610 @@
...
+    def __call__(self, sample, sample_posterior=False, deterministic: bool = True, return_dict: bool = True):
+        posterior = self.encode(sample, deterministic=deterministic, return_dict=return_dict)
+        if sample_posterior:
+            rng = self.make_rng("gaussian")
+            hidden_states = posterior.latent_dist.sample(rng)
+        else:
+            hidden_states = posterior.latent_dist.mode()
+        hidden_states = self.decode(hidden_states, return_dict=return_dict).sample
+
+        sample = jnp.transpose(hidden_states, (0, 3, 1, 2))
```
Comment (pcuenca) [inline src/diffusers/models/vae_flax.py:605]: Would this work with `pmap`?

File: src/diffusers/models/vae_flax.py [hunk 4]
```diff
@@ -0,0 +1,610 @@
...
+        posterior = DiagonalGaussianDistribution(moments)
+
+        if not return_dict:
+            return (posterior,)
+
+        return FlaxAutoencoderKLOutput(latent_dist=posterior)
+
+    def decode(self, latents, deterministic: bool = True, return_dict: bool = True):
+        if latents.shape[-1] != self.config.latent_channels:
+            latents = jnp.transpose(latents, (0, 2, 3, 1))
```
Comment (pcuenca) [inline src/diffusers/models/vae_flax.py:586]: Would this work with `pmap`?

File: src/diffusers/models/vae_flax.py [hunk 5]
```diff
@@ -0,0 +1,610 @@
...
+        sample_shape = (1, self.in_channels, self.sample_size, self.sample_size)
+        sample = jnp.zeros(sample_shape, dtype=jnp.float32)
+
+        params_rng, dropout_rng = jax.random.split(rng)
+        rngs = {"params": params_rng, "dropout": dropout_rng}
+
+        return self.init(rngs, sample)["params"]
+
+    def encode(self, sample, deterministic: bool = True, return_dict: bool = True):
+        sample = jnp.transpose(sample, (0, 2, 3, 1))
```
Comment (pcuenca) [inline src/diffusers/models/vae_flax.py:573]: Would this work with `pmap`? We might need to prepare the data externally and avoid the transpose here.

**Review by pcuenca — COMMENTED** [review]
Looks great. I'd suggest to test it with `pmap` if you haven't done it already, to ensure it works in multiple devices in jitted mode as well as in a single device.

File: src/diffusers/models/vae_flax.py [hunk 6]
```diff
@@ -0,0 +1,610 @@
...
+
+        if not return_dict:
+            return (hidden_states,)
+
+        return FlaxDecoderOutput(sample=hidden_states)
+
+    def __call__(self, sample, sample_posterior=False, deterministic: bool = True, return_dict: bool = True):
+        posterior = self.encode(sample, deterministic=deterministic, return_dict=return_dict)
+        if sample_posterior:
+            rng = self.make_rng("gaussian")
```
Comment (pcuenca) [inline src/diffusers/models/vae_flax.py:599]: I think this may require the root key `"gaussian"` to be passed in `init_weights()` (and also in `apply()` when we need to sample).

File: src/diffusers/models/vae_flax.py [hunk 7]
```diff
@@ -0,0 +1,610 @@
...
+            dtype=self.dtype,
+        )
+
+    def init_weights(self, rng: jax.random.PRNGKey) -> FrozenDict:
+        # init input tensors
+        sample_shape = (1, self.in_channels, self.sample_size, self.sample_size)
+        sample = jnp.zeros(sample_shape, dtype=jnp.float32)
+
+        params_rng, dropout_rng = jax.random.split(rng)
+        rngs = {"params": params_rng, "dropout": dropout_rng}
```
Comment (pcuenca) [inline src/diffusers/models/vae_flax.py:568]: Maybe replace `"dropout"` with `"gaussian"`.

File: src/diffusers/models/vae_flax.py (hunk 3 above)
Comment (patrickvonplaten) [reply src/diffusers/models/vae_flax.py:605]: Will check it! But think since `batch_dim` (`=0`) is untouched there shouldn't be a problem

File: src/diffusers/models/vae_flax.py (hunk 5 above)
Comment (patil-suraj) [reply src/diffusers/models/vae_flax.py:573]: Should work fine with `pmap`.

File: src/diffusers/models/vae_flax.py (hunk 3 above)
Comment (patil-suraj) [reply src/diffusers/models/vae_flax.py:605]: actually this should go inside `decode` method, because we call that method  in the pipeline.

**Review by patil-suraj — COMMENTED** [review]
Looks great! Just left some nits to align the names  with PyTorch.

Two things to be addressed before merging:
- Use lists for downsamplers and unsamplers, same way we do in PT. This will make auto-conversion possible.
- `transpose` the `output` in the `decode` method as that method is used in the pipeline.


For follow-up PR:
- We should move the blocks in its own file`unet_blocks_flax.py`
- Make sure all module names are identical to it's PT counterpart.

File: src/diffusers/models/vae_flax.py [hunk 8]
```diff
@@ -0,0 +1,610 @@
...
+        )
+
+    def __call__(self, hidden_states):
+        pad = ((0, 0), (0, 1), (0, 1), (0, 0))  # pad height and width dim
+        hidden_states = jnp.pad(hidden_states, pad_width=pad)
+        hidden_states = self.conv(hidden_states)
+        return hidden_states
+
+
+class ResnetBlock(nn.Module):
```
[the remark this answers is not in the document]
Comment (pcuenca) [reply src/diffusers/models/vae_flax.py:89]: Actually, there's a very similar `FlaxResnetBlock2D` block here: https://github.com/huggingface/diffusers/blob/main/src/diffusers/models/resnet_flax.py#L50

We should try to merge both implementations.

File: src/diffusers/models/vae_flax.py [hunk 9]
```diff
@@ -0,0 +1,610 @@
...
+                in_channels=in_channels,
+                out_channels=self.out_channels,
+                dropout_prob=self.dropout,
+                dtype=self.dtype,
+            )
+            resnets.append(res_block)
+        self.resnets = resnets
+
+        if self.add_downsample:
+            self.downsample = Downsample(self.out_channels, dtype=self.dtype)
```
[the remark this answers is not in the document]
Comment (patrickvonplaten) [reply src/diffusers/models/vae_flax.py:227]: fixing in a follow-up PR
